### ISSUE TITLE
Alternatives to MLE

### ISSUE THREAD

**Issue by harrysouthworth** [JuliaStats/GLM.jl#45]
I'm a complete noob to Julia. Please be patient with me whilst I find my feet.

For binomial GLMs, separation is a common problem in some areas of application. The general "solution" is MAP estimation, or penalized MLE. The methods of Firth and of Gelman et al work well (but ridge estimation would also do).

In R, these are implemented in various different functions in various different packages: brglm::brglm, arm::bayesglm. Both of these use fairly straightforward modifications to the IRLS algorithm. (Firth's method had a couple of previous incarnations using different fitting algorithms in the logistf and brlr packages.)

So far as I understand the Julia GLM package, it implements MLE. To avoid multiple future packages implementing alternatives such as those mentioned above, a general interface for penalized MLE would be useful, with predefined options for at least Firth's method.

Is such an interface reasonable, or does the required flexibility make it more sensible to implement separate fitting methods explicitly elsewhere? If total flexibility is unreasonable, can Firth's method be offered as an option (even SAS has it!)

Thanks,
Harry

http://biomet.oxfordjournals.org/content/80/1/27.short
http://www.stat.columbia.edu/~gelman/research/published/priors11.pdf

**Comment by Nosferican** [comment]
I took a look at Gelman et al. (2008) and it seems feasible. I could imagine something like
```
model = glm(formula::StatsModels.Formula,
            data::DataFrames.AbstractDataFrame,
            distribution::Distributions.Distribution;
            link::GLM.Link,
            priors::AbstractVector{<: Distributions.Distribution})
```
The default could probably be Ridge regression using standard normal distribution priors for all coefficients except the intercept after variable standardization. I rewrote IRLS for dense matrices using O'Leary (1990) IRLS QR Newton which is one of the methods in this package if not mistaken so I could give it a shot... Any help would be welcomed.
